rfcs/20181016-optimizer-unification.md

```diff
@@ -0,0 +1,398 @@
+
+
+### II - Add a from_config class method on every optimizer (only needs a single implementation on the base class):
+
+```python
+optimizer = Adadelta.from_config(config)
+```
+
+This method is already present on the Model class and every Layer class. This method is required for Keras compatibility.
+
+**Args:**
+- config: A dictionary, containing the same keys as what gets returned by `get_config`.
+
+**Returns:**
+- An optimizer instance with the desired configuration, effectively a clone of the original optimizer (minus its state, i.e. its weight values).
+
+
+### III - Add a get_weights method on every optimizer (only needs a single implementation on the base class):
+
+```python
+optimizer.get_weights()
+```
+
+This method is already present on the Model class and every Layer class.
+
+**Returns:**
+- A flat list of Numpy arrays, in deterministic order, where each array represents the value of an internal weight of the optimizer (such as the momentum of a model weight).
+
+
+### IV - Add a set_weights method on every optimizer (only needs a single implementation on the base class):
+
+```python
+optimizer.set_weights(weights)
+```
+
+This method is already present on the Model class and every Layer class. This method is required for Keras compatibility.
+
+**Args:**
+- weights: A flat list of Numpy arrays, in deterministic order, same as returned by get_weights. Note that since the optimizer creates its internal weights to match the set of weights it is trying to optimize, set_weights would only match get_weights when the set of weights being optimized is equivalent. E.g.:
+
+```python
+optimizer = Adadelta()
+_ = optimizer.get_weights()  # returns empty list since the optimizer has no weights at that point
+model.compile(optimizer=optimizer, loss=loss) # Weights are created here
+weights = optimizer.get_weights()  # Returns a list of numpy arrays
+optimizer.set_weights(weights)  # This works!
+
+# This will not work since this optimizer would have a different set weight
+different_model.optimizer.set_weights(weights)
+```
+
+Note: if the optimizer has been called on more than a single set of weights, we should disable `get_weights` and `set_weights` since their meaning would be ambiguous.
+
+
+### V - Make all optimizer hyperparameters accessible via attributes (they currently aren’t retrievable):
+
+```python
+optimizer = Adadelta(learning_rate=0.2)
+optimizer.learning_rate  # returns learning rate tensor
+```
+
+This should generally work for any numerical parameter that can be passed to the constructor.
+
+
+### VI - Make the following work on every optimizer, in both eager and graph modes:
+
+```python
+optimizer = Adadelta(learning_rate=0.2)
```

Comment:
Needs some clarification on whether the old methods of using a tensor as learning rate, is still supported, 
e.g. `Adadelta(learning_rate=tf.train.linear_cosine_decay(...))`. 
In this case `optimizer.learning_rate=0.1` should throw an error.

Reply:
It is